File: 백준/Platinum/3015. 오아시스 재결합/오아시스 재결합.cc
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define X first
#define Y second

int n;
stack<pair<int, int>> S;
ll ans;

int main(void) {
    cin>>n;
    
    while(n--) {
        int h;
        cin>>h;
        int cnt = 1; // var for count
        
        while(!S.empty()&&S.top().X<=h) { // if top of stack height is less than input height
            ans+=S.top().Y; // add count
            if(S.top().X==h) cnt+=S.top().Y; // if input same height raise count's value
            S.pop();
        }
        if(!S.empty()) ans++; // if stack isn't empty, it means that the height higher than input height is left in stack. so add count
        S.push({h,cnt});
    }
    
    cout<<ans;
    return 0;
}

File: 백준/Platinum/3015. 오아시스 재결합/오아시스 재결합.py
import sys
r = sys.stdin.readline
sys.setrecursionlimit(10**6)

def solution():
    n = int(r().rstrip())
    stack = []
    res = 0
    
    for _ in range(n):
        h = int(r().rstrip())
        cnt = 1
        
        while stack and stack[-1][0] <= h:
            height, cnt = stack.pop()
            if height == h:
                res += cnt
                cnt += 1
            elif height < h:
                res += cnt
                cnt = 1
        if stack: res += 1
        stack.append((h, cnt))
    print(res)
    
solution()